hbase-client/src/main/java/org/apache/hadoop/hbase/client/ImmutableScan.java

```diff
@@ -0,0 +1,477 @@
+    throw new IllegalStateException("ImmutableScan does not allow access to setSmall");
+  }
+
+  @Override
+  public Scan setAttribute(String name, byte[] value) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setAttribute");
+  }
+
+  @Override
+  public Scan setId(String id) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setId");
+  }
+
+  @Override
+  public Scan setAuthorizations(Authorizations authorizations) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setAuthorizations");
+  }
+
+  @Override
+  public Scan setACL(Map<String, Permission> perms) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setACL");
+  }
+
+  @Override
+  public Scan setACL(String user, Permission perms) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setACL");
+  }
+
+  @Override
+  public Scan setConsistency(Consistency consistency) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setConsistency");
+  }
+
+  @Override
+  public Scan setReplicaId(int Id) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setReplicaId");
+  }
+
+  @Override
+  public Scan setIsolationLevel(IsolationLevel level) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setIsolationLevel");
+  }
+
+  @Override
+  public Scan setPriority(int priority) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setPriority");
+  }
+
+  @Override
+  public Scan setScanMetricsEnabled(final boolean enabled) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setScanMetricsEnabled");
+  }
+
+  @Override
+  @Deprecated
+  public Scan setAsyncPrefetch(boolean asyncPrefetch) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setAsyncPrefetch");
+  }
+
+  @Override
+  public Scan setLimit(int limit) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setLimit");
+  }
+
+  @Override
+  public Scan setOneRowLimit() {
+    throw new IllegalStateException("ImmutableScan does not allow access to setOneRowLimit");
+  }
+
+  @Override
+  public Scan setReadType(ReadType readType) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setReadType");
+  }
+
+  @Override
+  Scan setMvccReadPoint(long mvccReadPoint) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setMvccReadPoint");
+  }
+
+  @Override
+  Scan resetMvccReadPoint() {
+    throw new IllegalStateException("ImmutableScan does not allow access to resetMvccReadPoint");
+  }
+
+  @Override
+  public Scan setNeedCursorResult(boolean needCursorResult) {
+    throw new IllegalStateException("ImmutableScan does not allow access to setNeedCursorResult");
+  }
+
+  @Override
+  public long getMaxResultSize() {
+    return super.getMaxResultSize();
```

Comment:
Do not need to implement these methods?

Reply:
Yes, no need but the reason why I prefer to keep them is because when any dev introduces new public method in `Scan`, they can realize (with IDE's help) that other methods are overridden in `ImmutableScan` and if the return value is Collection or something worth converting to Unmodifiable, they can do it.